### ISSUE TITLE
Support LZMA or LZO compression

### ISSUE THREAD

**Issue by jpap** [dpc/rdedup#44]
Looks like deflate is the only scheme supported.  Would be nice to see if we get further savings using LZMA and/or LZO, selectable at the time the repo is created.

**Comment by dpc** [comment]
Fine with me. This would depend on #39

**Comment by dpc** [comment]
I have just landed configurable compression. Adding new compression is now as easy as adding some structs in `lib/src/compression.rs` and couple a new enum variant here and there.

**Comment by dpc** [comment]
@jpap : Do you think we should we keep this issue open or close it?

BTW. Did you do any comparison between time of compression, time of decompression, repository size?

If you run `rdedup -vvvv` it should dump quite a bit of timing data. At least on writing path, compression itself included: https://github.com/dpc/rdedup/blob/master/lib/src/chunk_processor.rs#L58

**Comment by jpap** [comment]
Let's close this issue; if you want to discuss timings, we can continue to do so here or in a new issue as you feel.

Based on the results below, it might be worth considering bzip2 as well; it is likely to have significantly faster runtime compared to xz with better compression than deflate.  I'll create a new issue; it looks like the bzip2 Crate offers an API that is very similar to the flate2 and shouldn't be hard to implement.

As a quick test, I tar'ed up my rdedup git repo, which came out to be 49684480 bytes.

Xz gives almost 12% gain over deflate:
* deflate 23627942 bytes
* xz 20820720 bytes

If I'm reading the verbose logs right -- I've summed the "compression" timings on the chunk-processing, Xz is over 6x slower:
* deflate compression time: 3.473 sec
* xz compression time: 27.342 sec

Raw data follows below.

```
Deflate >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

Jun 07 20:13:04.162 INFO Opening write handle
Jun 07 20:13:04.163 INFO Writing data, name: test-1
Jun 07 20:13:06.022 INFO time report, name: input-reader, tx: 1.758624743, input: 0.097827112
Jun 07 20:13:06.022 INFO time report, name: chunker, tx: 1.529060371, rx-and-chunking: 0.32755627
Jun 07 20:13:06.057 INFO time report, name: chunker, rx-and-chunking: 1.866983104, tx: 0.006296838
Jun 07 20:13:06.058 INFO time report, name: chunker, rx-and-chunking: 0.983645709, tx: 0.00003658
Jun 07 20:13:06.061 INFO time report, name: index-processor, digest-rx: 0.98701643, spawn-chunker: 0.000027378, sorting-recv-create: 0.000001147
Jun 07 20:13:06.062 INFO time report, name: index-processor, digest-rx: 1.880887859, spawn-chunker: 0.001420583, writer-tx: 0.000001578, sorting-recv-create: 0.000000824
Jun 07 20:13:06.062 INFO time report, name: index-processor, digest-rx: 1.895761701, spawn-chunker: 0.000113463, sorting-recv-create: 0.000062581, writer-tx: 0.000001114
Jun 07 20:13:06.062 INFO time report, name: chunk-processing, compress: 0.8089616, tx-writer: 0.523511466, processing: 0.473125722, encrypt: 0.068348557, rx: 0.012404198, tx-digest: 0.007866352
Jun 07 20:13:06.062 INFO time report, name: chunk-processing, compress: 0.91766401, tx-writer: 0.464977851, processing: 0.427340844, encrypt: 0.074808025, rx: 0.007972647, tx-digest: 0.003665262
Jun 07 20:13:06.062 INFO time report, name: chunk-processing, compress: 0.899597971, tx-writer: 0.499755623, processing: 0.394932205, encrypt: 0.078179117, rx: 0.020906824, tx-digest: 0.003574795
Jun 07 20:13:06.062 INFO time report, name: chunk-processing, compress: 0.847687328, tx-writer: 0.533086571, processing: 0.430584213, encrypt: 0.071793514, rx: 0.011045328, tx-digest: 0.002613591
Jun 07 20:13:06.101 INFO time report, name: chunk-writer, fsync: 0.891197624, write: 0.629337141, rx: 0.39445033, processing: 0.019057574
Jun 07 20:13:06.101 INFO time report, name: chunk-writer, fsync: 0.923042728, write: 0.561985691, rx: 0.431871147, processing: 0.016983613
Jun 07 20:13:06.101 INFO time report, name: chunk-writer, fsync: 0.968309796, write: 0.569693586, rx: 0.372553073, processing: 0.02384758
Jun 07 20:13:06.101 INFO time report, name: chunk-writer, fsync: 0.800009107, write: 0.697880213, rx: 0.396445958, processing: 0.038265617
Jun 07 20:13:06.101 INFO time report, name: chunk-writer, fsync: 0.893295277, write: 0.589623787, rx: 0.395743763, processing: 0.055348456
Jun 07 20:13:06.102 INFO time report, name: chunk-writer, fsync: 0.837680153, write: 0.624842161, rx: 0.408336454, processing: 0.064301303
Jun 07 20:13:06.111 INFO time report, name: chunk-writer, fsync: 0.861919508, write: 0.631726255, rx: 0.417074081, processing: 0.032253294
Jun 07 20:13:06.122 INFO time report, name: chunk-writer, fsync: 0.832473166, write: 0.726318573, rx: 0.377495261, processing: 0.018404345
Jun 07 20:13:06.131 INFO time report, name: chunk-writer, fsync: 0.822115509, write: 0.691393787, rx: 0.441924899, processing: 0.00802045
Jun 07 20:13:06.131 INFO time report, name: chunk-writer, fsync: 0.824897891, write: 0.665531547, rx: 0.434470864, processing: 0.038685848
Jun 07 20:13:06.131 INFO time report, name: chunk-writer, fsync: 0.966590361, write: 0.618520518, rx: 0.354965796, processing: 0.024550966
Jun 07 20:13:06.131 INFO time report, name: chunk-writer, fsync: 0.913963304, write: 0.633308114, rx: 0.380850066, processing: 0.035321526
Jun 07 20:13:06.131 INFO time report, name: chunk-writer, fsync: 0.866045793, write: 0.668261196, rx: 0.401988284, processing: 0.028270237
Jun 07 20:13:06.132 INFO time report, name: chunk-writer, fsync: 0.853728467, write: 0.697177764, rx: 0.383980013, processing: 0.028502458
Jun 07 20:13:06.132 INFO time report, name: chunk-writer, fsync: 0.900983906, write: 0.576569927, rx: 0.437423304, processing: 0.046163139
Jun 07 20:13:06.132 INFO time report, name: chunk-writer, fsync: 0.806615515, write: 0.691258332, rx: 0.438778774, processing: 0.027186854
517 new chunks
23627942 new bytes
```

```
Xz >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

Jun 07 20:15:49.799 INFO Opening write handle
Jun 07 20:15:49.800 INFO Writing data, name: test-1
Jun 07 20:15:56.720 INFO time report, name: input-reader, tx: 6.741480686, input: 0.178384733
Jun 07 20:15:56.838 INFO time report, name: chunker, tx: 6.768805476, rx-and-chunking: 0.266356117
Jun 07 20:15:56.919 INFO time report, name: chunker, rx-and-chunking: 7.001728231, tx: 0.026998797
Jun 07 20:15:56.919 INFO time report, name: chunker, rx-and-chunking: 2.5463038239999998, tx: 0.000026974
Jun 07 20:15:56.919 INFO time report, name: index-processor, digest-rx: 2.546762139, spawn-chunker: 0.000123038, sorting-recv-create: 0.000001657
Jun 07 20:15:56.920 INFO time report, name: index-processor, digest-rx: 7.029459989, spawn-chunker: 0.002208471, writer-tx: 0.00000518, sorting-recv-create: 0.000001765
Jun 07 20:15:56.920 INFO time report, name: index-processor, digest-rx: 7.118701381, spawn-chunker: 0.000060292, sorting-recv-create: 0.000022318, writer-tx: 0.000001058
Jun 07 20:15:56.920 INFO time report, name: chunk-processing, compress: 6.810640965, processing: 0.144308702, encrypt: 0.065758603, rx: 0.056051054, tx-writer: 0.030791973, tx-digest: 0.011609988
Jun 07 20:15:56.920 INFO time report, name: chunk-processing, compress: 6.855891422, processing: 0.136716373, encrypt: 0.076376856, rx: 0.032059613, tx-writer: 0.011867831, tx-digest: 0.006482708
Jun 07 20:15:56.920 INFO time report, name: chunk-processing, compress: 6.817403411, processing: 0.167141719, encrypt: 0.084348953, rx: 0.028048322, tx-writer: 0.016666336, tx-digest: 0.00586322
Jun 07 20:15:56.920 INFO time report, name: chunk-processing, compress: 6.857600667, processing: 0.143371697, encrypt: 0.079592796, rx: 0.016968214, tx-writer: 0.016211817, tx-digest: 0.006027174
Jun 07 20:15:56.921 INFO time report, name: chunk-writer, rx: 6.833033586, fsync: 0.259731123, write: 0.024828766, processing: 0.001804593
Jun 07 20:15:56.921 INFO time report, name: chunk-writer, rx: 6.8600234239999995, fsync: 0.238452981, write: 0.019221083, processing: 0.00165361
Jun 07 20:15:56.921 INFO time report, name: chunk-writer, rx: 6.848617344, fsync: 0.245910109, write: 0.022830219, processing: 0.002138963
Jun 07 20:15:56.921 INFO time report, name: chunk-writer, rx: 6.837317352, fsync: 0.257308276, write: 0.022472366, processing: 0.002407265
Jun 07 20:15:56.921 INFO time report, name: chunk-writer, rx: 6.816818366, fsync: 0.277262042, write: 0.019776663, processing: 0.005640785
Jun 07 20:15:56.921 INFO time report, name: chunk-writer, rx: 6.815026862, fsync: 0.285069576, write: 0.017449257, processing: 0.001990045
Jun 07 20:15:56.921 INFO time report, name: chunk-writer, rx: 6.807267254, fsync: 0.288651886, write: 0.021268123, processing: 0.002377522
Jun 07 20:15:56.922 INFO time report, name: chunk-writer, rx: 6.7891386449999995, fsync: 0.298925368, write: 0.029446197, processing: 0.002105026
Jun 07 20:15:56.922 INFO time report, name: chunk-writer, rx: 6.829211647, fsync: 0.273641769, write: 0.013464716, processing: 0.003497689
Jun 07 20:15:56.922 INFO time report, name: chunk-writer, rx: 6.820526704, fsync: 0.269476249, write: 0.020020458, processing: 0.009634067
Jun 07 20:15:56.922 INFO time report, name: chunk-writer, rx: 6.830706991, fsync: 0.273268397, write: 0.014014204, processing: 0.001651918
Jun 07 20:15:56.924 INFO time report, name: chunk-writer, rx: 6.836978725, fsync: 0.263929554, write: 0.019781716, processing: 0.002157542
Jun 07 20:15:56.929 INFO time report, name: chunk-writer, rx: 6.824252745, fsync: 0.283799542, write: 0.017558984, processing: 0.001811432
Jun 07 20:15:56.929 INFO time report, name: chunk-writer, rx: 6.821571474, fsync: 0.28545944, write: 0.018337854, processing: 0.001889182
Jun 07 20:15:56.929 INFO time report, name: chunk-writer, rx: 6.849799792, fsync: 0.250941587, write: 0.025024647, processing: 0.001696949
Jun 07 20:15:56.929 INFO time report, name: chunk-writer, rx: 6.873228773, fsync: 0.237086082, write: 0.015356778, processing: 0.001748893
517 new chunks
20820720 new bytes
```

**Comment by dpc** [comment]
All right. Thanks again!
